src/mqt/problemsolver/partialcompiler/Partial_QAOA.py

```diff
@@ -0,0 +1,192 @@
+        assert len(qcs_problem_uncompiled) == len(qcs_mix_uncompiled)
+        qc_composed = qc_prep.copy()
+        for i in range(len(qcs_problem_uncompiled)):
+            qc_composed.compose(qcs_problem_uncompiled[i], inplace=True)
+            qc_composed.compose(qcs_mix_uncompiled[i], inplace=True)
+
+        offline_mapped_qc = transpile(
+            qc_composed, coupling_map=self.coupling_map, basis_gates=IBM_GATES, optimization_level=3
+        )
+
+        layout = offline_mapped_qc._layout.initial_layout
+        mapping = []
+        for elem in layout.get_virtual_bits():
+            if elem.register.name == "ancilla":
+                pass
+            else:
+                mapping.append(layout.get_virtual_bits()[elem])
+        return mapping
+
+    def get_partially_compiled_circuit_without_online_edges(
+        self, consider_mapping: bool = True
+    ) -> tuple[QuantumCircuit, list[QuantumCircuit], list[QuantumCircuit]]:
+        """
+        Return the state preparation circuit and lists of problem and mixer circuits with the option to consider
+        the mapping with all returned circuits are compiled to native gates and optionally mapped to the device.
+        """
+
+        self.mapping = self.determine_mapping()
+
+        qc_prep, qcs_problem_uncompiled, qcs_mix_uncompiled = self.get_uncompiled_circuit()
+
+        mapping_fct = self.compile_with_mapping if consider_mapping else self.compile_without_mapping
+        qc_prep_compiled = mapping_fct(qc_prep)
+        qcs_problem_compiled = []
+        qcs_mixer_compiled = []
+        for i in range(len(qcs_problem_uncompiled)):
+            qcs_problem_compiled.append(mapping_fct(qcs_problem_uncompiled[i]))
+            qcs_mixer_compiled.append(mapping_fct(qcs_mix_uncompiled[i]))
+
+        return qc_prep_compiled, qcs_problem_compiled, qcs_mixer_compiled
+
+    def compile_without_mapping(self, qc: QuantumCircuit, opt_level: int = 3) -> QuantumCircuit:
+        return transpile(
+            qc,
+            basis_gates=IBM_GATES,
+            optimization_level=opt_level,
+        )
+
+    def compile_with_mapping(self, qc: QuantumCircuit, opt_level: int = 3) -> QuantumCircuit:
+        return transpile(
+            qc,
+            basis_gates=IBM_GATES,
+            initial_layout=self.mapping,
+            coupling_map=self.coupling_map,
+            layout_method="trivial",
+            optimization_level=opt_level,
+        )
+
+    def get_uncompiled_fully_composed_circuit(self) -> QuantumCircuit:
+        """
+        Returns the fully composed circuit without any compilation.
+        """
+        qc_prep, qcs_problem, qcs_mix = self.get_uncompiled_circuit(include_online_edges=True)
+        qc = qc_prep
+        for i in range(len(qcs_problem)):
+            qc.compose(qcs_problem[i], inplace=True)
+            qc.compose(qcs_mix[i], inplace=True)
+        return qc
```

Comment:
I guess it would make more sense to add a boolean flag to the `get_uncompiled_circuit` method that indicates whether the result should be a single quantum circuit or a this trifecta of circuits.

You can even type hint that appropriately using https://adamj.eu/tech/2021/05/29/python-type-hints-how-to-use-overload/

Reply:
I am not sure whether I understand your thought correctly. The `overload` declarator would be used if a parameter accepts multiple input types which would lead to different return types. In the case of a flag indicating whether the whole circuit shall be returned, the type would always be the same and its value would determine the return type. So a `Union` for the return type would suffice?

Reply:
You can use `Literal[True]` and `Literal[False]` to distinguish the two.

Reply:
See, e.g., https://stackoverflow.com/questions/59359943/python-how-to-write-typing-overload-decorator-for-bool-arguments-by-value

Reply:
Added that. Thanks for the hints :)